optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/ConstraintCollectors.java

```diff
@@ -1966,6 +1968,121 @@ public void remove(Key key, Value value) {
                                 ConstraintCollectors::throwOnKeyConflict, TreeMap::new)));
     }
 
+    // Consecutive
+
+    /**
+     * Creates a constraint collector that returns {@link ConsecutiveData} about the first
+     * fact.
+     *
+     * For instance ${@code [Shift slot=1] [Shift slot=2] [Shift slot=4] [Shift slot=6]}
+     * returns the following infomation
+     *
+     * ${@code
+     * Consecutive Lengths: 2 1 1
+     * Break Lengths: 1 2
+     * Consecutive Items [[Shift slot=1] [Shift slot=2]], [[Shift slot=4]] [[Shift slot=6]]
+     * }
+     * 
+     * @param indexMap Maps the fact to its position in the sequence
+     * @param <A> type of the first mapped fact
+     * @return never null
+     */
+    public static <A> UniConstraintCollector<A, ConsecutiveSetTree<A, Integer, Integer>, ConsecutiveData<A, Integer>>
+            consecutive(ToIntFunction<A> indexMap) {
+        return new DefaultUniConstraintCollector<>(
+                () -> new ConsecutiveSetTree<A, Integer, Integer>((Class<? extends A>) Object.class,
+                        indexMap::applyAsInt,
+                        (Integer a, Integer b) -> b - a, 1, 0),
+                (acc, a) -> {
+                    acc.add(a);
+                    return () -> {
+                        acc.remove(a);
+                    };
+                },
+                ConsecutiveSetTree::getConsecutiveData);
+    }
+
+    /**
+     * As defined by {@link #consecutive(ToIntFunction)} (Function, Function, BinaryOperator)}.
+     *
+     * @param resultMap Maps both facts to an item in the sequence
+     * @param indexMap Maps the item to its position in the sequence
+     * @param <A> type of the first mapped fact
+     * @param <B> type of the second mapped fact
+     * @param <Result> type of item in the sequence
+     * @return never null
+     */
+    public static <A, B, Result>
+            BiConstraintCollector<A, B, ConsecutiveSetTree<Result, Integer, Integer>, ConsecutiveData<Result, Integer>>
+            consecutive(BiFunction<A, B, Result> resultMap, ToIntFunction<Result> indexMap) {
```

Comment:
Let's discuss later which of these methods we really want to expose as the public API.